runtime/lua/vim/ui/clipboard/osc52.lua

```diff
@@ -1,60 +1,71 @@
 local M = {}
 
-function M.copy(lines)
-  local s = table.concat(lines, '\n')
-  io.stdout:write(string.format('\027]52;;%s\027\\', vim.base64.encode(s)))
+-- The default format string of the OSC 52 sequence. This will be used unless
+-- the caller provides a separate format string (e.g. one obtained via terminfo
+-- or XTGETTCAP)
+local osc52 = '\027]52;%s;%s\027\\'
+
+function M.copy(reg, format)
+  local clipboard = reg == '+' and 'c' or 's'
+  return function(lines)
+    local s = table.concat(lines, '\n')
+    io.stdout:write(string.format(format or osc52, clipboard, vim.base64.encode(s)))
+  end
 end
 
-function M.paste()
-  local contents = nil
-  local id = vim.api.nvim_create_autocmd('TermResponse', {
-    callback = function(args)
-      local resp = args.data ---@type string
-      local encoded = resp:match('\027%]52;%w?;([A-Za-z0-9+/=]*)')
-      if encoded then
-        contents = vim.base64.decode(encoded)
-        return true
-      end
-    end,
-  })
-
-  io.stdout:write('\027]52;;?\027\\')
-
-  local ok, res
-
-  -- Wait 1s first for terminals that respond quickly
-  ok, res = vim.wait(1000, function()
-    return contents ~= nil
-  end)
-
-  if res == -1 then
-    -- If no response was received after 1s, print a message and keep waiting
-    vim.api.nvim_echo(
-      { { 'Waiting for OSC 52 response from the terminal. Press Ctrl-C to interrupt...' } },
-      false,
-      {}
-    )
-    ok, res = vim.wait(9000, function()
+function M.paste(reg, format)
+  local clipboard = reg == '+' and 'c' or 's'
+  return function()
+    local contents = nil
+    local id = vim.api.nvim_create_autocmd('TermResponse', {
+      callback = function(args)
+        local resp = args.data ---@type string
+        local encoded = resp:match('\027%]52;%w?;([A-Za-z0-9+/=]*)')
```

Comment:
Note: it's _technically_ possible for some pathological terminal to use some other sequence than the standard OSC 52 sequence, in which case this pattern won't match. For now we assume that no such terminals exist and that if they do they just won't work. If this assumption ever needs to be revisited then this line will need to become a bit smarter (we would need to convert the `format` parameter from a `string.format` format to a `string.match` pattern).

Reply:
I went ahead and nipped this in the bud by ignoring Ms capabilities that don't match `OSC 52`: https://github.com/neovim/neovim/pull/26064/files#diff-776616fe246164a7a47bb8dcca8edfcdc1d00c3a3525badb93547dde5341ffbcR18-R22